fix: apply serializer to `Error` instance for thrown snapshot

hi-ogawa opened the pull request:
### Description

Closes https://github.com/vitest-dev/vitest/issues/3655

As explained in https://github.com/vitest-dev/vitest/issues/3655#issuecomment-1784471527, currently vitest hard-codes specific behavior to extract `Error.message` for `Error` instance when it's snapshot-ed via `toThrowErrorMatching...`, which prevents error from being applied with user-provided serializer.
This PR removes this special behavior, so that `Error` instance's snapshot is a consistent regardless of whether its snapshote-ed via `toMatchInlineSnapshot` or `toThrowErrorMatchingInlineSnapshot`.

Due to the removal of `Error.message` extraction, the default snapshot format changes in a following way

```tsx
// before
  expect(() => {
    throw new Error('omega')
  }).toThrowErrorMatchingInlineSnapshot(`"ometa"`)

// after
  expect(() => {
    throw new Error('omega')
  }).toThrowErrorMatchingInlineSnapshot(`[Error: omega]`)
```


<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

<!-- You can also add additional context here -->

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

sheremet-va commented:
This looks fine to me. We already break _evey_ snapshot with default values in Vitets 1.0, so might as well.

hi-ogawa commented:
Hi, thanks for confirming. Good to know that the breaking change is acceptable now.
I'll continue with this PR to cleanup some code and also adjust some test cases.

sheremet-va commented:
Would also be nice to add this to the migration from Jest guide. Jest uses `error.message` currently.

hi-ogawa commented:
> Would also be nice to add this to the migration from Jest guide. Jest uses `error.message` currently.

Good point. I didn't think about jest compatibility. I'll take a look at the documentation.

hi-ogawa commented:
I updated the documentation to reflect the change in this PR.
I wasn't sure which, but instead of `guide/migration.md`, I updated `guide/snapshot.md` since it has a dedicated "Difference from Jest" section.

hi-ogawa commented on docs/guide/snapshot.md:
````diff
@@ -237,5 +237,32 @@ exports[`toThrowErrorMatchingSnapshot: hint 1`] = `"error"`;
+
+```js
+test('snapshot', () => {
+  //
+  // in Jest
+  //
+
+  expect(new Error('error')).toMatchInlineSnapshot(`[Error: error]`)
+
+  // Jest snapshots `Error.message` for `Error` instance
+  expect(() => {
+    throw new Error('error')
+  }).toThrowErrorMatchingInlineSnapshot(`"error"`)
````
I verified this behavior in https://stackblitz.com/edit/jest-example-t1qnq1?file=sum.test.js

<details><summary>reveal screenshot</summary>

![image](https://github.com/vitest-dev/vitest/assets/4232207/0cd51d10-73ae-4e0a-8207-e5367ae7b7e8)


</details>
